shapes: do not read_back entire shape to get aliases uids

voodoos opened the pull request:
Fully reducing the shapes of large modules (and all their components) is an expensive process. Merlin, when trying to jump to a module, used to do that to get the uid of the module, and thus it's location. This wasteful reduction slowed the query and we introduced "weak_reduction" which does not reduce a module's components when all we are interested in is the module's uid itself.

In #12508, we redesigned this feature along with @Ekdohibs and @gasche in the form of the `reduce_for_uid` function that doesn't leak incorrect shapes anymore. This function performs weak reduction and returns only the resulting uid. When looking for a module alias it returns the list of the aliases uids and the aliased module uid:

```
module X = struct end
module A_1 = X
...
module A_N = A_N-1
```

`reduce_for_uid 'path to A_N'` returns `[uid_of_A_N; ...; uid_of_A_1; uid_of_X]`. 

This information allows Merlin to traverse aliases and jump to the actual definition of X, but keep aliasing information which is useful for other use cases like `occurrences`.

To return this list I used a call  to `read_back` before looking at the shape description to get the head aliases. This was, of course, a mistake since it will perform full reduction of the module's body. This lead to blow up when indexing large codebases and slower locate queries when jumping to an aliased module.

This PR fixes that by only forcing step by step the shape reduction to get the head aliases. 

@Octachron I think the performance issue I described should not have a large impact during the partial indexation that we are introducing in 5.2 because it only reduce shapes locally. It would take a very pathological compilation unit for the issue to manifest itself in a significant way. Still, it might be better to have this fix in the 5.2 branch.

voodoos committed:
shapes: do not read_back entire shape to get aliases uids

gasche commented on typing/shape_reduce.ml:
```diff
@@ -271,6 +271,22 @@ end) = struct
+  (* When interested only of in the uids of aliased modules we do not read_back
+    the entire shape of the module, just enough to unroll the chain of aliases.
+  *)
+  let read_back_aliases_uids env (nf : nf) =
```
My feeling is that the code is correct, the comment is somewhat redundant (I think it is clear from the code), and the naming is slightly wrong, as you are not reading back (to generate a term). I would call this `reduce_aliases_for_uid` for example.

voodoos replied:
My reasoning for the name was that functions of the "reduce"  take a shape as an input while "read_back" functions take a normal form as input. I am ok with both and change it in [4c3f43f](https://github.com/ocaml/ocaml/pull/13001/commits/4c3f43faf46872fbe1315e8bd24b9d69a8f0090a)

gasche commented on typing/shape_reduce.ml:
```diff
@@ -271,6 +271,22 @@ end) = struct
+  let read_back_aliases_uids env (nf : nf) =
+    let force (Thunk (local_env, t)) =
+      reduce_ { env with local_env } t
+    in
```
I think that `let force env (Trunk (local_env, t)) = ...` should be a function defined more globally, to be shared with the other copies of this definition.

voodoos replied:
Done in [4c3f43f](https://github.com/ocaml/ocaml/pull/13001/commits/4c3f43faf46872fbe1315e8bd24b9d69a8f0090a)

gasche commented on typing/shape_reduce.ml:
```diff
@@ -323,7 +331,7 @@ end) = struct
       Unresolved (read_back env nf)
```
This occurrence of `read_back` is also a risk for performance, right? (Did we discuss making this a `unit -> shape` function to let the user force the read-back on demand?)

voodoos replied:
These results are marshaled and written to the `cmt` files, I guess that should dissuade us to put closures in them ? We could make it configurable via a function parameter when calling `reduce_for_uid` ?

edit: it would not make much sense to make it configurable since the main usage for this case is to let an external tool finish the reduction later. My intuition is that since we are stuck on a Proj (Comp_unit _, _) the read_back should not be expensive because the actual module shape is not loaded anyway.

gasche replied:
You are right that there is no obviously good choice here. I am ok with forgetting about that case and focusing on the other one, that you identified as being an issue in practice, for the present PR.

gasche commented on typing/shape_reduce.ml:
```diff
@@ -323,7 +331,7 @@ end) = struct
     else match nf with
       | { desc = NAlias _; approximated = false; _ } ->
-          get_aliases_uids (read_back env nf)
+          read_back_aliases_uids env nf
```
I wonder if the code would look nicer if we had

```ocaml
type result =
   ...
  | Resolved_alias of uid.t * result
```

so that we would track alias chains also in the "approximated" and other cases.

voodoos committed:
Rename function and extract the `force` utility.

voodoos committed:
New result type : `Resolved_alias of Uid.t * result`

voodoos committed:
Add tests illustrating issue with shapes that
should be marked approximated but are not

The tests also illustrate that we can get aliases to an
"approximated" module.

voodoos commented:
@gasche I applied your type-change suggestion in [83108d3](https://github.com/ocaml/ocaml/pull/13001/commits/83108d3dc66fc70ffd4cb580880fa2317c5fbae2).

While adding new tests to show that we can return aliases to approximate modules I noticed that we didn't mark the resulting shape as approximated for values coming from a first-class-module. I did that in [19a778f](https://github.com/ocaml/ocaml/pull/13001/commits/19a778f26f6aeaf1ff032979fc955f730264878b). The reduction is "successful", but there is no uid.

gasche commented on testsuite/tests/shape-index/index_aliases.reference:
```diff
@@ -21,13 +32,22 @@ Resolved: Index_aliases.1 :
 Index_aliases.6: F' (File "index_aliases.ml", line 22, characters 7-9)
 Index_aliases.3: t (File "index_aliases.ml", line 21, characters 23-24)
+Index_aliases.11: s (File "index_aliases.ml", line 31, characters 39-40)
 Index_aliases.0: t (File "index_aliases.ml", line 18, characters 23-24)
```
There are many more names listed here now, do you agree that this is a good change? (Or is it a regression in some respect?)

voodoos replied:
That's expected, this list reflects the fact that I added new definitions in the module. The ordering is a bit awkward, but no regression here.

gasche commented on typing/shape_reduce.ml:
```diff
@@ -197,8 +198,11 @@ end) = struct
               end
+          | NLeaf ->
+              (* Happens when projecting from a first-class module *)
+              approx_nf (nored ())
```
Do we want a similar case in the `App(f, args)` block above, if `f.desc` is a `NLeaf`? (Can you test this?)

voodoos replied:
That's not clear. I added a case that would be impacted by such a change in [d682f39](https://github.com/ocaml/ocaml/pull/13001/commits/d682f39950ff7d06711d0ef14603010d0bbd0556). 

If in `module FArg = X.F (Arg)`, `X` is actually an instance of a first class module, so the shape of `FArg` is indeed unsolvable. But when reducing the shape of `Farg` "for uid", we don't really care about that since we want the uid of the binding we made when writing `FArg = ... `. We do care however if we reduce the shape of `Farg.t`, which should return `approximated no uid`  (but right now returns `no uid` without the approximated mention.).

gasche replied:
I would find it more natural to have the same behavior for reduction of functor application and structure projection. If you don't have a strong preference, I would do the same on both sides.

voodoos replied:
I pushed in [b9ba1b9](https://github.com/ocaml/ocaml/pull/13001/commits/b9ba1b99643f6c8ac301d3c9e112cae059eeba9f) a test for `Farg.t` and added a case to effectively mark it as approximated instead of returning an `Internal_error_no_uid`.

I think that maybe `approximated` is not a good name or the best concept: it actually only occurs when accessing items inside first-class-modules ? (But I won't have time to clean this now, I will be away from this Friday until the beginning of April.)

gasche replied:
Hm, the new code does not look like what I would expect: I don't think the `NApp _` case here is correct. If you don't want to work on this now, I would suggest removing this part of the code from the present PR (it is not really in scope anyway).

More details: projecting out of a NLeaf is the sign of a problem, because NLeaf should not be the shape of a structure. On the other hand, NApp can naturally occur on open terms -- there is an external compilation unit under it. What I had in mind was to add a `NLeaf` case in the `NApp (f, args)` case above, that is, where we currently have `| _ ->` line 190. Just like you did in the projection case, in the case where this catch-all is actually a leaf we know that an approximation took place.

voodoos replied:
> I would find it more natural to have the same behavior for reduction of functor application and structure projection. If you don't have a strong preference, I would do the same on both sides.

I knew you would say that :-) 

But if I change the behavior for application it will break occurrences of modules that are new bindings for the result of functor applications that are approximated. For example, the indexed occurrence of the `FArg` module would be: 

```diff
- Resolved: Index_aliases.26 :
+ Approximated: Index_aliases.26 :
  FArg (File "index_aliases.ml", line 48, characters 5-9)
 ```
 
But in that case `Index_aliases.26` is really the identifier of the definition of `FArg`, it's not an approximation.

If we want to keep symmetry until we find a better solution then I suggest we remove the other case: having some uids registered with an `Internal_error_missing_uid` instead of an `Approximated None` does not have an impact on the actual useful occurrences while having changes such as the one in the diff for uid `Index_aliases.26` will prevent some useful occurrences to be gathered.

If that's ok with you I will rollback the changes i made to this PR that are not related to its initial purpose: drastic performance improvement in the presence of aliases. The approximation issues are not critical right now and we can tackle them when we will have more time.

voodoos replied:
So our messages interleaved twice, I think we agree on rewinding back to the original intent of this PR, I will do that.

gasche reviewed (commented):
Thanks! This looks nicer. See two questions inline.

voodoos commented:
@gasche I removed the fix for approximation. I think we do want to have the `NLeaf` cases mark the shape as approximated in both the `App` and `Proj` case. But we need first improve `reduce_for_uid`'s lazyness (or have another way to check for approximation):

Right now, when looking for the uid of `M` when `M` is defined by `module M = F (X)`, reducing the shape will reduce the application. However we don't need to perform that reduction, since the uid of M's definition is the uid of the  `module M = ...` binding, independently of the right-handside.

Reducing the right hand-side might result in an approximated shape, but this is not something we care about when reducing the shape for the uid of M.

gasche commented:
My current (and ever-changing) understanding is that there are two different needs:
1. For the compiler, you want a type of "partially reduced shapes" that is computed and then serialized in the `cmt_ident_occurrences` field of `cmt` files. We want to work now instead of having the tools do more work, so we want an "eager" reduction.
2. For the tools, you want a type of "head normal form" that gives the minimal amount of information you need while doing the minimal amount of work, in particular the reduction should "lazy".

Currently we are trying to use a single `result` type for this, and the PR here and the discussion suggests that this might not be the right choice. If we separated the two, we could use for (2) a version that has some laziness and is not serializable.

For (1), I wonder why we are not "just" using the `Shape.t` value that results from strong reduction -- instead of storing the result of `reduce_for_uid`, the result of `reduce`. You report that earlier versions tried to store the shapes of all values and it lead to a blowup in cmt size, but were those the input shapes or the strongly-reduced shapes? Or do we maybe know that we will only ever nead the head uid of those shapes, not the whole shape, computing below the head normal form is a waste?

For (2), I think that we should use a representation that allows on-demand computation, for example:

```
| Resolved_alias of uid.t * result Lazy.t
```

voodoos committed:
Add changelog entry for #13001

voodoos commented:
> My current (and ever-changing) understanding is that there are two different needs:
> 
>     1. For the compiler, you want a type of "partially reduced shapes" that is computed and then serialized in the `cmt_ident_occurrences` field of `cmt` files. We want to work now instead of having the tools do more work, so we want an "eager" reduction.
> 
>     2. For the tools, you want a type of "head normal form" that gives the minimal amount of information you need while doing the minimal amount of work, in particular the reduction should "lazy".

I don't think that this is correct. In (1) reducing the shapes to build the `cmt_ident_occurrences` table we have the same need as for (2): do the minimal amount of work to get the minimal amount of information: the uid of the definition. However, these reductions might stay incomplete since in (1) we don't load other modules' cmt files to respect separate compilation.

> For (1), I wonder why we are not "just" using the Shape.t value that results from strong reduction -- instead of storing the result of reduce_for_uid, the result of reduce. 

Apart from the fact that, right now, tools are only interested in the result uid, `reduce` performs unconditional `read_back`s which can be costly.

> You report that earlier versions tried to store the shapes of all values and it lead to a blowup in cmt size, but were those the input shapes or the strongly-reduced shapes?

That was with the "original" shapes stored in the typing environment.

Octachron commented on typing/shape_reduce.mli:
```diff
@@ -18,10 +18,10 @@
 type result =
   | Resolved of Shape.Uid.t (** Shape reduction succeeded and a uid was found *)
-  | Resolved_alias of Shape.Uid.t list (** Reduction led to an alias chain *)
+  | Resolved_alias of Shape.Uid.t * result (** Reduction led to an alias *)
```
If possible I would rather avoid changing the interface for OCaml 5.2.0, but I am not sure if we are still aiming for 5.2.0 for this PR.

voodoos replied:
> If possible I would rather avoid changing the interface for OCaml 5.2.0

Is it because of the already released alpha and beta versions with bumped magic numbers ? Since the `result` type was actually introduced in 5.2 itself, it seems like it could be better to change it right now... 

I can also easily rollback to the first iteration on this PR that fixes the performance issue without changing the type if that is required.

> I am not sure if we are still aiming for 5.2.0 for this PR

The discussion diverged to other unclear parts of the shape reduction, but I think this PR's original scope is smaller than that and mostly uncontroversial: removing extraneous work done in some cases, with an agreed-on fix (stepped read-back). We do need to re-think shape reduction and it's handling of uids at some point, but that's out-of-scope for that pr. What do you think @gasche ?

Octachron replied:
It is more a question of API stability for shape clients. In particular, this change would require a new beta and a patch to odoc. This is still ok-ish, if the PR converges this week, but this is starting to get late in the release cycle for this kind of changes.

voodoos replied:
I understand, thanks for the details, I can rollback the API changes if it make things simpler. Now that you mention odoc, it reminds me that the performance fix might actually be important for their usage of the shapes. (We discussed it a few weeks ago with @panglesd  and @Julow, they perform similar actions as we do in Merlin to identify definitions' uids.)

panglesd replied:
`odoc` would indeed would benefit from the improvements of this PR! (The "render source code" feature of `odoc` is probably very inefficient by itself, but we will work on it when it is usable by drivers.)

Regarding API changes, we haven't yet released a 5.2 compatible version of odoc, so I think that for us it is still fine to include it.

Octachron replied:
Ok, let's have a beta 2 and I will amend my patch for odoc for 5.2.

Ekdohibs reviewed (approved):
The changes look good, and will avoid useless reductions indeed.
I don't have a strong opinion about the API changes; the new API seems better than the previous one, however. In any case, I'd say that at least the basic version of this PR should be merged for 5.2, in order to have the performance improvements.

Octachron reviewed (approved):
I am green-ticking on behalf of @Ekdohibs

gasche commented:
I won't have the time to look at this again in the next 10 days, but I trust @Ekdohibs' review and would be happy to approve on her behalf. @Octachron, can you say "yes" on improving the interface?

gasche commented on Changes:
```diff
@@ -118,6 +118,10 @@ _______________
 
+- #13001: do not read_back entire shapes to get aliases' uids when building the
+  usages index
+  (Ulysse Gérard, review by Gabriel Scherer)
```
Let's also add Nathanaëlle.

Octachron replied:
And let's move the entry in the 5.2 section too.

gasche reviewed (approved):
Same here.

voodoos committed:
Add @Ekdohibs to #13001 changelog entry

Octachron commented:
And for people reading the conversation in a linear way, yes I agree with the change of API.

voodoos committed:
Move changelog entry to section 5.2.0

Octachron commented:
Cherry-picked to 5.2.0 in 89301a2ca .
